Automattic/wp-calypso
# Themes: Add verticals description to SSR'd HTML

This PR:
* Enables the data layer middleware on the server (yes, works fine :slightly_smiling_face: )
* Updates the theme filters schema to reflect the API
* Add a `requestThemeFilters` action and a number of related selectors (plus unit tests)...
* ... which are used by the `themes` section's controller to fetch themes before server-side rendering the showcase
* Changes the `ThemeShowcase` component to use the right data for the `description` `meta` tag.

To test:
* Restart Calypso: `DEBUG="calypso:themes" make run`
* While logged out, verify in Page Source that there's a specific `description` `meta` in the following cases (you need to land on these pages directly):
  * Vertical:  http://calypso.localhost:3000/themes/photography
  * Single Filter: http://calypso.localhost:3000/themes/filter/minimal
  * Tier: http://calypso.localhost:3000/themes/premium
  * Generic: http://calypso.localhost:3000/themes/
* Note that not all verticals and filters have a description; in those cases, we fall back to the next lower level (i.e. no verticals -> filter; neither verticals nor filters -> tier).
* Multiple filters are equivalent to no filters at all (i.e. the next lower level is used).
* Verify that after the first hit, theme filters are cached (watch out for `found theme filters in cache` in your node console).

---

(Original description:)

Background (from Slack):

> @dmsnell I have a conceptual question about the data/HTTP middleware. As you know, the `themeFilters` reducer is implemented to use that. For server-side rendering, we require data fetched by the corresponding middleware. With other network fetches implemented as `requestSomething` thunks, we usually chain `.then( next )` in the corresponding `page.js` controller in order to hold execution until we've received the required data. Q is, how would I go about achieving the same effect with the data/HTTP mw, where all consumer-facing actions are synchronous?

FYI @dmsnell @blowery @gwwar 

Edit: props @blowery for the idea how to have the router's middleware wait for theme filters to be fetched.

https://github.com/Automattic/wp-calypso/pull/10988 might be a reasonable future step to simplify data caching.
